src/tests/ftest/telemetry/engine_events.py

```diff
@@ -0,0 +1,353 @@
+                           f"After = {events_last_event_ts_1[rank]}")
+                    errors.append(msg)
+                    result = (f"Rank {rank}: {events_last_event_ts_0[rank]} -> "
+                              f"{events_last_event_ts_1[rank]} Fail (Value changed)")
+                else:
+                    result = (f"Rank {rank}: {events_last_event_ts_0[rank]} -> "
+                              f"{events_last_event_ts_1[rank]} Pass (Value unchanged)")
+            else:
+                if events_last_event_ts_0[rank] >= events_last_event_ts_1[rank]:
+                    msg = (f"No increase detected in engine_events_last_events_ts for rank {rank}! "
+                           f"Before = {events_last_event_ts_0[rank]}; "
+                           f"After = {events_last_event_ts_1[rank]}")
+                    errors.append(msg)
+                    result = (f"Rank {rank}: {events_last_event_ts_0[rank]} -> "
+                              f"{events_last_event_ts_1[rank]} Fail (Value decreased/unchanged)")
+                else:
+                    result = (f"Rank {rank}: {events_last_event_ts_0[rank]} -> "
+                              f"{events_last_event_ts_1[rank]} Pass (Value increased)")
+            events_last_event_ts_results.append(result)
+
+    def test_engine_events(self):
+        """Test engine-related events after an engine is restarted.
+
+        1. Check that all ranks have started.
+        2. Gather initial data for the required metrics for all the server nodes.
+        3. Stop a rank.
+        4. Verify the desired rank has stopped successfully.
+        5. Restart the stopped rank after several seconds.
+        6. Verify the desired rank restarted successfully.
+        7. Gather final data for required metrics with the same command as in step 2.
+        8. Compare the before and after values.
+
+        :avocado: tags=all,full_regression
+        :avocado: tags=hw,medium
+        :avocado: tags=telemetry
+        :avocado: tags=EngineEvents,test_engine_events
+        """
+        # 1. Check that all ranks have started.
+        self.log_step("Check that all ranks have started.")
+        dmg_command = self.get_dmg_command()
+        if not check_system_query_status(data=dmg_command.system_query()):
+            self.fail("Engine failure detected at the beginning of the test!")
+
+        # 2. Gather initial data for the required metrics for all the server nodes.
+        self.log_step("Gather initial data for the required metrics for all the server nodes.")
+        rank_count = self.server_managers[0].engines
+        telemetry_before = self.collect_telemetry(rank_count=rank_count)
+
+        # 3. Stop a rank.
+        restart_rank = rank_count - 1
+        self.log_step(f"Stop rank {restart_rank} and wait for a while for RPCs to circulate")
+        dmg_command.system_stop(ranks=str(restart_rank))
+        # Wait for the RPCs to go around the ranks.
+        time.sleep(40)
```

Comment:
We should not be using sleeps in test code where it can be avoided.  Is there a something that can be queried from the server or its log that would indicate that it has progressed far enough for the test to continue?

Reply:
I updated the code to periodically check `engine_events_dead_ranks` of the ranks that aren't stopped. If the RPCs circulate, at least one of the ranks becomes 1.